File: r/sparklabs/batch/sales/dataframe/customers_never_ordered.R
source("r/sparklabs/batch/sales/dataframe/utils.R")

customer_ids_with_orders <- distinct(select(orders, "customer_id"))
result <- except(customers, join(customers, customer_ids_with_orders, customers$customer_id == customer_ids_with_orders$customer_id, "inner"))
showDF(result, truncate = FALSE)

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataframe/CustomersNeverOrdered.scala
package io.devorbit.sparklabs.batch.sales.dataframe

import org.apache.spark.sql.functions._

object CustomersNeverOrdered extends App {
  Utils.customers
    .join(Utils.orders, Seq("customer_id"), "left_anti")
    .show(false)
//  Thread.sleep(100000000)

}
